rfcs/2021-10-29-8621-framing-and-codecs-sinks.md

```diff
@@ -0,0 +1,126 @@
+
+## Cross cutting concerns
+
+The codec work is very interweaved with the ongoing [schema work](https://github.com/vectordotdev/vector/pull/9388). It introduces the concept of "mappers" which understand how to transform an event at runtime based on the schema from its source and destination within Vector's topology.
+
+We want to hand over the reshaping responsibilities to the schema, and provide a transition layer to support existing reshaping capabilities of sinks until schema mappers are implemented.
+
+Another related current work is `StandardEncoding` which has been introduced in [#9215](https://github.com/vectordotdev/vector/pull/9215). It provides a consistent set of codec implementations aims and therefore aligns with our goals. However, it does not separate the reshaping responsibilities.
+
+## Scope
+
+### In scope
+
+- Common support structures for encoding/framing
+- Set of commonly used encoding (text, JSON, ...) and framing (newline-delimited, octet-counting, ...) methods
+- Strategy for integrating encoding/framing with sinks
+
+### Out of scope
+
+- Integration with schema system (automatic transformation of event fields depending)
+
+## Pain
+
+Internally, we don't share implementations for encoding strategies that are very commonly used (e.g. JSON, newline-delimited). This leads to duplicated effort and missing guidance for handling encoding in sinks since no canonical implementation exists.
+
+Externally, users are restricted by the available custom set of encodings implemented by the sink. There does not exist any mental model for consistent encoding/framing options for event data that applies to all sinks.
+
+## Proposal
+
+### User Experience
+
+Users should be able to set `encoding` and `framing` options on sinks, analogously to the `decoding` and `framing` options on sources. These options uniformly control how the event _payload_ is encoded. This distinction is important, as encoding for the sink specific _protocol_ and the event _payload_ are separate concerns. The payload should still be encoded according to the sink's protocol, and the sink should provide additional options if there are multiple protocols to choose from, e.g. under a `protocol` key.
+
+The fields containing event transformations in sinks on the current `encoding` options (`schema`, `only_fields`, `except_fields`, `timestamp_format`) should be moved to a dedicated option, e.g. `schema` or `transform`. Thus, this would introduce a breaking change in configurations. However, migration would be relatively straight-forward by nesting these options under the new key.
```

Comment:
The one question that I think needs to be answered here is: will the schema work land before this happens?  If it doesn't, then we probably need a way handle parsing these out of the existing `encoding` configuration field.  We could deprecate setting them via `encoding` at the same time, but we would have to provide a deprecation window regardless.

Reply:
Yeah, I would suggest we continue to support the current config interface for now and focus on organizing the internal implementation for now. Then once both this work and the schema work is further along we can tackle changing the external API with a bit more confidence.

Reply:
Sounds like a good plan 👍 Should be able to support existing config options without carrying over too much baggage.

Reply:
I came up with a wrapper that is able to apply the existing transformation behavior to the new encoding options while being fully backwards compatible to the old sink-specific encoding options: https://github.com/vectordotdev/vector/pull/10253.

For usage, have a look at https://github.com/vectordotdev/vector/pull/10684.

Reply:
Clarified in 21f9c36d82fdc5cbe8a299f7c7677a35bfd4ae24.